File: src/main/java/org/agentic/flink/tools/mcp/McpClient.java
package org.agentic.flink.tools.mcp;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicLong;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Minimal JSON-RPC 2.0 client for MCP servers.
 *
 * <p>Implements just enough of the protocol to {@code initialize}, {@code tools/list} and
 * {@code tools/call} — which is enough to turn any MCP server into a set of
 * {@link org.agentic.flink.tools.ToolExecutor}s. Notifications and server-initiated
 * messages are ignored (they are not relevant to the tool-execution flow).
 *
 * <p>Not thread-safe: each operator task should hold its own client instance, constructed once
 * in {@code RichFunction.open()}.
 */
public final class McpClient implements AutoCloseable {

  private static final Logger LOG = LoggerFactory.getLogger(McpClient.class);
  private static final String PROTOCOL_VERSION = "2024-11-05";

  private final McpServerSpec spec;
  private final ObjectMapper mapper = new ObjectMapper();
  private final AtomicLong nextId = new AtomicLong(1);

  // STDIO-specific state
  private Process process;
  private BufferedWriter stdin;
  private BufferedReader stdout;

  // HTTP-specific state
  private HttpClient http;

  private volatile boolean initialized = false;

  public McpClient(McpServerSpec spec) {
    this.spec = spec;
  }

  /** Connect and complete the MCP {@code initialize} handshake. Idempotent. */
  public synchronized void initialize() throws IOException {
    if (initialized) return;
    switch (spec.getTransport()) {
      case STDIO:
        startStdio();
        break;
      case HTTP:
        startHttp();
        break;
    }

    ObjectNode params = mapper.createObjectNode();
    params.put("protocolVersion", PROTOCOL_VERSION);
    params.set("capabilities", mapper.createObjectNode());
    ObjectNode info = mapper.createObjectNode();
    info.put("name", "agentic-flink");
    info.put("version", "1.0.0");
    params.set("clientInfo", info);

    JsonNode result = rpc("initialize", params);
    LOG.info(
        "MCP server '{}' initialized: serverInfo={}",
        spec.getName(), result == null ? null : result.path("serverInfo"));
    initialized = true;
  }

  /** List the tools exposed by the connected server. */
  public List<McpToolMetadata> listTools() throws IOException {
    ensureInitialized();
    JsonNode result = rpc("tools/list", mapper.createObjectNode());
    List<McpToolMetadata> tools = new ArrayList<>();
    if (result == null) return tools;
    JsonNode arr = result.path("tools");
    if (!arr.isArray()) return tools;
    for (JsonNode t : arr) {
      String name = t.path("name").asText();
      String description = t.path("description").asText("");
      Map<String, Object> schema = new HashMap<>();
      JsonNode schemaNode = t.path("inputSchema");
      if (schemaNode.isObject()) {
        Iterator<Map.Entry<String, JsonNode>> it = schemaNode.fields();
        while (it.hasNext()) {
          Map.Entry<String, JsonNode> e = it.next();
          schema.put(e.getKey(), mapper.convertValue(e.getValue(), Object.class));
        }
      }
      tools.add(new McpToolMetadata(name, description, schema));
    }
    return tools;
  }

  /** Invoke a tool by name with the given argument map. Returns the tool's content payload. */
  public Object callTool(String toolName, Map<String, Object> arguments) throws IOException {
    ensureInitialized();
    ObjectNode params = mapper.createObjectNode();
    params.put("name", toolName);
    params.set("arguments", mapper.valueToTree(arguments == null ? Map.of() : arguments));
    JsonNode result = rpc("tools/call", params);
    if (result == null) return null;
    // MCP returns { content: [{ type: "text", text: "..." }, ...], isError: bool }
    boolean isError = result.path("isError").asBoolean(false);
    JsonNode content = result.path("content");
    StringBuilder text = new StringBuilder();
    if (content.isArray()) {
      for (JsonNode part : content) {
        if ("text".equals(part.path("type").asText())) {
          if (text.length() > 0) text.append('\n');
          text.append(part.path("text").asText());
        }
      }
    }
    String body = text.toString();
    if (isError) {
      throw new IOException("MCP tool '" + toolName + "' returned error: " + body);
    }
    return body;
  }

  @Override
  public synchronized void close() {
    try {
      if (stdin != null) stdin.close();
    } catch (IOException ignored) {
    }
    try {
      if (stdout != null) stdout.close();
    } catch (IOException ignored) {
    }
    if (process != null) {
      process.destroy();
      try {
        if (!process.waitFor(2, java.util.concurrent.TimeUnit.SECONDS)) {
          process.destroyForcibly();
        }
      } catch (InterruptedException e) {
        Thread.currentThread().interrupt();
        process.destroyForcibly();
      }
    }
    initialized = false;
  }

  // ---- internals ----

  private void ensureInitialized() throws IOException {
    if (!initialized) {
      initialize();
    }
  }

  private void startStdio() throws IOException {
    ProcessBuilder pb = new ProcessBuilder(spec.getCommand());
    pb.environment().putAll(spec.getEnv());
    pb.redirectErrorStream(false);
    process = pb.start();
    stdin = new BufferedWriter(new OutputStreamWriter(process.getOutputStream(), StandardCharsets.UTF_8));
    stdout = new BufferedReader(new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8));
  }

  private void startHttp() {
    http = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(10)).build();
  }

  private JsonNode rpc(String method, ObjectNode params) throws IOException {
    long id = nextId.getAndIncrement();
    ObjectNode req = mapper.createObjectNode();
    req.put("jsonrpc", "2.0");
    req.put("id", id);
    req.put("method", method);
    req.set("params", params);

    String payload = mapper.writeValueAsString(req);
    JsonNode response;
    if (spec.getTransport() == McpServerSpec.Transport.STDIO) {
      stdin.write(payload);
      stdin.write('\n');
      stdin.flush();
      String line = readResponseLine(id);
      response = mapper.readTree(line);
    } else {
      response = httpRpc(payload);
    }
    if (response.has("error")) {
      throw new IOException("MCP rpc error: " + response.get("error").toString());
    }
    return response.get("result");
  }

  private String readResponseLine(long expectedId) throws IOException {
    // The server may interleave notifications (no id) with our response. Skip until we see the
    // matching id.
    String line;
    while ((line = stdout.readLine()) != null) {
      JsonNode parsed;
      try {
        parsed = mapper.readTree(line);
      } catch (Exception e) {
        LOG.debug("Skipping non-JSON line from MCP server: {}", line);
        continue;
      }
      if (parsed.has("id") && parsed.get("id").asLong() == expectedId) {
        return line;
      }
      // Notification or unrelated — log and continue.
      LOG.debug("MCP notification ignored: {}", line);
    }
    throw new IOException("MCP server closed before responding to id=" + expectedId);
  }

  private JsonNode httpRpc(String payload) throws IOException {
    try {
      HttpRequest.Builder b =
          HttpRequest.newBuilder()
              .uri(URI.create(spec.getUrl()))
              .timeout(Duration.ofSeconds(30))
              .header("Content-Type", "application/json")
              .POST(HttpRequest.BodyPublishers.ofString(payload));
      for (Map.Entry<String, String> h : spec.getHeaders().entrySet()) {
        b.header(h.getKey(), h.getValue());
      }
      HttpResponse<String> resp = http.send(b.build(), HttpResponse.BodyHandlers.ofString());
      if (resp.statusCode() / 100 != 2) {
        throw new IOException("HTTP " + resp.statusCode() + ": " + resp.body());
      }
      return mapper.readTree(resp.body());
    } catch (InterruptedException ie) {
      Thread.currentThread().interrupt();
      throw new IOException(ie);
    }
  }
}

File: agentic-clj/src/agentic/mcp_client.clj
(ns agentic.mcp-client
  "MCP stdio client — connect to a Model Context Protocol server over newline-delimited JSON-RPC 2.0
   and surface its tools in a `ToolRegistry`, so an agent can call external MCP tool servers. The
   Clojure mirror of jagentic-core's McpStdioClient and pyagentic's McpClient. No third-party MCP
   dependency — just a BufferedReader/Writer pair (real process via java.lang.ProcessBuilder, or any
   piped streams for tests).

   Protocol: `initialize` -> notification `notifications/initialized` -> `tools/list` -> `tools/call`.
   A `tools/call` result's text is the join (with \"\\n\") of its `content` blocks where type==\"text\"."
  (:require [clojure.data.json :as json]
            [clojure.string :as str]
            [agentic.tools :as tools])
  (:import [java.io BufferedReader Writer InputStreamReader OutputStreamWriter]
           [java.lang ProcessBuilder ProcessBuilder$Redirect]
           [java.nio.charset StandardCharsets]))

(def ^:private protocol-version "2024-11-05")

(defrecord McpClient [^BufferedReader reader ^Writer writer counter lock process specs])

(defn- write-line [^Writer wtr ^String json-str]
  (.write wtr json-str)
  (.write wtr "\n")
  (.flush wtr))

(defn- notification [client method params]
  (write-line (:writer client)
              (json/write-str {:jsonrpc "2.0" :method method :params (or params {})})))

(defn- request
  "Write a JSON-RPC request and block until the response with a matching id arrives, skipping
   notifications and non-JSON lines. Throws on a JSON-RPC error or premature EOF. Synchronized so
   concurrent callers don't interleave their reads."
  [client method params]
  (locking (:lock client)
    (let [id (swap! (:counter client) inc)
          ^Writer wtr (:writer client)
          ^BufferedReader rdr (:reader client)]
      (write-line wtr (json/write-str {:jsonrpc "2.0" :id id :method method :params (or params {})}))
      (loop []
        (let [line (.readLine rdr)]
          (when (nil? line)
            (throw (ex-info (str "MCP server closed before responding to " method) {:method method})))
          (if (str/blank? line)
            (recur)
            (let [node (try (json/read-str line) (catch Exception _ ::skip))]
              (cond
                (= node ::skip) (recur)
                (= (get node "id") id)
                (if-let [error (get node "error")]
                  (throw (ex-info (str "MCP error: " (json/write-str error)) {:error error :method method}))
                  node)
                :else (recur)))))))))

(defn- discover-specs
  "Run the initialize handshake and return the [{:name :description}] from tools/list."
  [client]
  (request client "initialize"
           {:protocolVersion protocol-version
            :capabilities {}
            :clientInfo {:name "agentic-clj" :version "0.1.0"}})
  (notification client "notifications/initialized" {})
  (let [listed (request client "tools/list" {})
        tools (get-in listed ["result" "tools"])]
    (vec (for [t (if (sequential? tools) tools [])]
           {:name (get t "name" "") :description (get t "description" "")}))))

(defn mcp-client-from-streams
  "Build a client over an already-open reader/writer pair and run the initialize handshake. The
   testable core constructor — works with any java.io.BufferedReader + java.io.Writer."
  [^BufferedReader reader ^Writer writer]
  (let [client (->McpClient reader writer (atom 0) (Object.) nil (atom []))
        specs (discover-specs client)]
    (reset! (:specs client) specs)
    client))

(defn mcp-client
  "Spawn an MCP server process via java.lang.ProcessBuilder (command is a vector like
   [\"python\" \"-m\" \"server\"]), wire its stdin/stdout, and run the handshake. Returns a client.
   `env` (optional map) is merged into the child environment. stderr is left to the server (not
   merged into stdout, so it can't corrupt the JSON-RPC stream)."
  ([command] (mcp-client command nil))
  ([command env]
   (let [^java.util.List cmd (vec (map str command))
         pb (ProcessBuilder. cmd)]
     (.redirectErrorStream pb false)
     (when (seq env)
       (let [pe (.environment pb)]
         (doseq [[k v] env] (.put pe (str k) (str v)))))
     (let [process (.start pb)
           writer (OutputStreamWriter. (.getOutputStream process) StandardCharsets/UTF_8)
           reader (BufferedReader. (InputStreamReader. (.getInputStream process) StandardCharsets/UTF_8))
           base (mcp-client-from-streams reader writer)]
       (assoc base :process process)))))

(defn list-tools
  "The discovered tool specs: [{:name :description}]."
  [client]
  @(:specs client))

(defn call-tool
  "Invoke an MCP tool and return the joined text of its result `content` text blocks."
  [client name args]
  (let [resp (request client "tools/call" {:name name :arguments (or args {})})
        content (get-in resp ["result" "content"])]
    (->> (if (sequential? content) content [])
         (filter #(= "text" (get % "type")))
         (map #(get % "text" ""))
         (str/join "\n"))))

(defn register
  "Register every discovered MCP tool into `reg` (ids `(str prefix name)`); returns `reg`."
  [client reg prefix]
  (let [p (or prefix "")]
    (doseq [{:keys [name description]} (list-tools client)]
      (tools/register reg (str p name) description
                      (fn [params] (call-tool client name params)))))
  reg)

(defn close
  "Best-effort shutdown: close the writer and, if a process was spawned, destroy + await it."
  [client]
  (try (.close ^Writer (:writer client)) (catch Exception _ nil))
  (when-let [^Process p (:process client)]
    (.destroy p)
    (try (.waitFor p) (catch InterruptedException _ (.interrupt (Thread/currentThread))))))
